[YUNIKORN-2886] update Spark operator documentation for YuniKorn integration (#474)
apache/yunikorn-site: docs/user_guide/workloads/run_spark.md

under the License.
-->

> ## Deploy with Spark operator and Helm

Author: Yeah, its better to align it.

:::note
Pre-requisites:
- This tutorial assumes YuniKorn is [installed](../../get_started/get_started.md) under the namespace `yunikorn`
- Use spark-operator version >= 2.0 to enable support for YuniKorn gang scheduling
:::

:::warning
This installation involves installing YuniKorn and Spark operator, which may take a few minutes to complete. To check the status we can use `kubectl get pods -n yunikorn` and `kubectl get pods -n spark-operator`
:::

### Install yunikorn
A simple script to install YuniKorn under the namespace `yunikorn`, refer to [Get Started](../../get_started/get_started.md) for more details.
```shell script
helm upgrade --install yunikorn yunikorn/yunikorn \
  --create-namespace \
  --namespace yunikorn
```

### Install spark operator

> We should install with the batch scheduler enabled and set the default scheduler to YuniKorn. It's optional to set the default scheduler to YuniKorn since you can specify it later on, but it's recommended to do so.

Author: Done.

under the License.
-->

> ## Deploy Spark job with Spark Operator and Helm

Reviewer: ```suggestion
## Run a Spark job with Spark Operator
```
Helm is standard tooling. Can omit in the title for brevity

:::note
Pre-requisites:
- This tutorial assumes YuniKorn is [installed](../../get_started/get_started.md) under the namespace `yunikorn`
- Use spark-operator version >= 2.0 to enable support for YuniKorn gang scheduling
:::

> :::warning

Reviewer: Consider changing it to `note` instead of `warning`. Spending a few minutes on installation doesn't seem to warrant a warning

Author: Thank you for pointing it out. I actually forget to delete it. It's not a problem this version of steps.

This installation involves installing YuniKorn and Spark operator, which may take a few minutes to complete. To check the status we can use `kubectl get pods -n yunikorn` and `kubectl get pods -n spark-operator`
:::

### Install YuniKorn

A simple script to install YuniKorn under the namespace `yunikorn`, refer to [Get Started](../../get_started/get_started.md) for more details.

```shell script
helm repo add yunikorn https://apache.github.io/yunikorn-release
helm repo update
helm install yunikorn yunikorn/yunikorn --create-namespace --namespace yunikorn
```

> ### Install spark operator

Reviewer: ```suggestion
### Install Spark Operator
```

We should install `spark-operator` with `controller.batchScheduler.enable=true` and set `controller.batchScheduler.default=yunikorn`. It's optional to set the default scheduler to YuniKorn since you can specify it later on, but it's recommended to do so.  

> Also, note that our total allocated memory is `memory + memoryOverhead + spark.executor.pyspark.memory`, which will further propagate to the `yunikorn.apache.org/task-groups` annotation for subsequent use. 

Author: once [spark-operator#2209](https://github.com/kubeflow/spark-operator/pull/2209) merged and released, it would be `memory + memoryOverhead + spark.executor.pyspark.memory + spark.memory.offHeap.size` opened a jira to track it: https://issues.apache.org/jira/browse/YUNIKORN-2919

> Also, note that our total allocated memory is `memory + memoryOverhead + spark.executor.pyspark.memory`, which will further propagate to the `yunikorn.apache.org/task-groups` annotation for subsequent use. 

Reviewer 2: > Also, note that our total allocated memory is `memory + memoryOverhead + spark.executor.pyspark.memory`, which will further propagate to the `yunikorn.apache.org/task-groups` annotation for subsequent use."

"our total allocated memory" is not accuacy, the `spark.executor.pyspark.memory` is only added to `spark-executor`'s tasks.

Driver = memory + memoryOverhead
Execturor = memory + memoryOverhead + spark.executor.pyspark.memory

Author: Done. for now it's clearer.

> Also, note that our total allocated memory is `Driver + Execturor`, where Driver = `memory + memoryOverhead` and Execturor = `memory + memoryOverhead + spark.executor.pyspark.memory`, which will further propagate to the `yunikorn.apache.org/task-groups` annotation for subsequent use. 

Reviewer 2: ```suggestion
Also, note that our total allocated memory is `Driver + Execturor`, where Driver = `memory + memoryOverhead` and Execturor = `memory + memoryOverhead + spark.executor.pyspark.memory`, which will further propagate to the `yunikorn.apache.org/task-groups` annotation. 
```

> Also, note that our total allocated memory is `Driver + Execturor`, where Driver = `memory + memoryOverhead` and Execturor = `memory + memoryOverhead + spark.executor.pyspark.memory`, which will further propagate to the `yunikorn.apache.org/task-groups` annotation. 

Reviewer: This part can use some rephrasing to be more clear. 
```suggestion
Also, note that the total requested memory for the Spark job is the sum of memory requested for the driver and that for all executors, where each is computed as below:
* Driver requested memory = `spark.driver.memory` + `spark.driver.memoryOverhead`
* Executor requested memory = `spark.executor.memory` + `spark.executor.memoryOverhead` + `spark.executor.pyspark.memory`
```
The requested memory for the driver and executors will propagate to the `yunikorn.apache.org/task-groups` annotation on the Spark pods.

helm repo update # (optional) update the helm repo cache to prvent install wrong version
```

```shell script
helm upgrade --install spark-operator spark-operator/spark-operator \
  --create-namespace \
  --namespace spark-operator \
  --set controller.batchScheduler.enable=true \
  --set controller.batchScheduler.default=yunikorn

> ```

Reviewer 2: I prefer to put all in one block. Similar to something in RayOperator setup:
- https://yunikorn.apache.org/docs/user_guide/workloads/run_ray_cluster#setup-a-kuberay-operator

```shell script
helm repo add spark-operator https://kubeflow.github.io/spark-operator
helm repo update 
helm upgrade --install spark-operator spark-operator/spark-operator \
  --create-namespace \
  --namespace spark-operator \
  --set controller.batchScheduler.enable=true \
  --set controller.batchScheduler.default=yunikorn
```

No need to add the 'optional' phrase.
And could we change the upgrade `helm upgrade` to `helm install`?  We are trying to install spark-operator here.

Author: Hi @chenyulin0719, I get your point. The best practice indeed is:
```
helm repo add spark-operator https://kubeflow.github.io/spark-operator
helm repo update 
helm install spark-operator spark-operator/spark-operator \
  --create-namespace \
  --namespace spark-operator \
  --set controller.batchScheduler.enable=true \
  --set controller.batchScheduler.default=yunikorn
```

> ### Create the example application

Reviewer: ```suggestion
### Create an example application
```
